src/modm/math/geometry/vector.hpp

```diff
@@ -11,141 +12,552 @@
-
-		const Matrix<T, 1, N>&
-		asTransposedMatrix() const;
-
-	public:
-		static inline uint8_t
-		getSize();
-
-		T coords[N];
-	};
-
-	template< typename T, uint8_t N >
-	struct detail::MakeSigned< Vector<T, N> >
-	{ using type = Vector< SignedType<T>, N >; };
-
-	template< typename T, uint8_t N >
-	struct detail::MakeUnsigned< Vector<T, N> >
-	{ using type = Vector< UnsignedType<T>, N >; };
-
-	template< typename T, uint8_t N >
-	struct detail::WideType< Vector<T, N> >
-	{ using type = Vector< WideType<T>, N >; };
-}
+		return Vector(
+			y() * other.z() - z() * other.y(),
+			z() * other.x() - x() * other.z(),
+			x() * other.y() - y() * other.x()
+		);
+	}
+
+	// various methods for the lengh of the vector
+
+	constexpr auto getLengthSquared() const {
+		// TODO treat the right CalcType
+		// - is always unsigned
+		// - Must fit std::pow(std::numeric_limits<T>::max(), 2) * N
+		using CalcType = WideType;
 
-#define IMPLEMENT_VECTOR_ACCESSOR2(a,b)			\
-		Vector<T, 2> a##b() const				\
-		{										\
-			return Vector<T, 2>(a, b);			\
-		}
+		CalcType sum{0};
 
-#define IMPLEMENT_VECTOR_ACCESSOR3(a, b, c)		\
-		Vector<T, 3> a##b##c() const			\
-		{										\
-			return Vector<T, 3>(a, b, c);		\
-		}
+		for(auto &v: *this)
+			sum += WideType(v) * WideType(v);
+		return sum;
+	}
 
-#define IMPLEMENT_VECTOR_ACCESSOR4(a, b, c, d)	\
-		Vector<T, 4> a##b##c##d() const			\
-		{										\
-			return Vector<T, 4>(a, b, c, d);	\
-		}
+	constexpr T getLength() const
+	{ return std::sqrt(getLengthSquared()); }
 
-#include "vector_impl.hpp"
+	constexpr T getLength() const requires std::integral<T>
+	{ return std::round(std::sqrt(getLengthSquared())); }
 
-#include "vector1.hpp"
-#include "vector2.hpp"
-#include "vector3.hpp"
-#include "vector4.hpp"
+	constexpr WideType getDistanceTo(const Vector& other) const
+	{ return (other - *this).getLength(); }
+
+	constexpr Vector& scale(float length){
+		operator*=( length / getLength() );
+		return *this;
+	}
+
+	constexpr Vector scaled(float length) const
+	{ return *this * (length / getLength()); }
+
+	// methods for float Vectors only
+
+	// Normalize length to 1
+	constexpr Vector& normalize()
+	requires std::floating_point<T> {
+		operator/=( getLength() );
+		return *this;
+	}
+
+	constexpr Vector normalized()
+	const requires std::floating_point<T>
+	{ return *this / getLength(); }
+
+
+	constexpr bool hasNan() const
+	requires std::floating_point<T>
+	{ return std::any_of(this->begin(), this->end(), [](T c){return std::isnan(c);}); }
+
+	constexpr bool hasInf() const
+	requires std::floating_point<T>
+	{ return std::any_of(this->begin(), this->end(), [](T c){return std::isinf(c);}); }
+
+	constexpr bool hasNan() const { return false; }
+
+	constexpr bool hasInf() const { return false; }
+
+	// methods for 2D only
+
+	constexpr Vector
+	operator~ () const
+	requires std::is_signed_v<T> && (N == 2)
+	{ return Vector(y(), -x()); }
+
+	constexpr float getAngle() const requires (N == 2)
```

Comment:
I've restricted some of the 2D methods to 2D vectors instead of generalizing them. Let's improve these things on demand, if ever.